addons/mail_client_extension/controllers/main.py

```diff
@@ -0,0 +1,323 @@
+        api_key = request.env['res.users.apikeys']._generate(scope, auth_message['name'])
+        return {'access_token': api_key }
+
+    def _get_auth_code_data(self, auth_code):
+        db_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        data, auth_code_signature = auth_code.split('.')
+        data = base64.b64decode(data)
+        auth_code_signature = base64.b64decode(auth_code_signature)
+        signature = hmac.new(db_secret.encode(), data, hashlib.sha256).hexdigest().encode()
+        if not hmac.compare_digest(auth_code_signature, signature):
+            return False, None
+
+        auth_message = json.loads(data)
+        # Check the expiration
+        if datetime.datetime.utcnow() - datetime.datetime.fromtimestamp(auth_message['timestamp']) > datetime.timedelta(minutes=3):
+            return False, None
+
+        return True, auth_message
+
+    # Using UTC explicitly in case of a distributed system where the generation and the signature verification do not
+    # necessarily happen on the same server
+    def _generate_auth_code(self, scope, name):
+        db_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        auth_dict = {
+            'scope': scope,
+            'name': name,
+            'timestamp': int(datetime.datetime.utcnow().timestamp()),  # <- elapsed time should be < 3 mins when verifying
+            'uid': request.uid,
+        }
+        auth_message = json.dumps(auth_dict, sort_keys=True).encode()
+        signature = hmac.new(db_secret.encode(), auth_message, hashlib.sha256).hexdigest().encode()
+        auth_code = "%s.%s" % (base64.b64encode(auth_message).decode(), base64.b64encode(signature).decode())
+        _logger.info('Auth code created - user %s, scope %s', request.env.user, scope)
+        return auth_code
+
+    def _iap_enrich(self, domain):
+        enriched_data = {}
+        try:
+            response = request.env['iap.enrich.api']\
+                .with_user(request.env.user)\
+                ._request_enrich({domain: domain}) # The key doesn't matter
+        except odoo.addons.iap.models.iap.InsufficientCreditError as ice:
+            _logger.info('Enrich request: failed because of credit')
+            enriched_data['enrichment_info'] = {'type': 'insufficient_credit', 'info': request.env['iap.account'].with_user(request.env.user).get_credits_url('reveal')}
+        except Exception as e:
+            enriched_data["enrichment_info"] = {'type': 'other', 'info': 'Unknown reason'}
+        else:
+            enriched_data = response.get(domain)
+            _logger.info('Enrich request: success')
+            if not enriched_data:
+                enriched_data = {'enrichment_info': {'type': 'no_data', 'info': 'The enrichment API found no data for the email provided.'}}
+        return enriched_data
+
+    def _iap_enrich_partner_from_response(self, env, iap_data, partner):
```

Comment:
Method seems not used anymore, remove?

Reply:
And also the dependency to `crm_iap_lead_enrich`?

Reply:
To remove? No, iap enrichment is used elsewhere (just above in fact, the `_iap_enrich` method which is called from `_create_company_from_iap`), it's just that this specific method is not used.

Reply:
Ouch, a pity to force IAP in this base module. Shouldn't that be done through glue modules?